source/agora/consensus/validation/Block.d

```diff
@@ -466,14 +470,31 @@ public bool isGenesisBlockValid (const ref Block genesis_block)
     return isGenesisBlockInvalidReason(genesis_block) is null;
 }
 
-/// Ditto but returns `bool`, only usable in unittests
+/// Ditto but returns `bool` and logs reason if fails, only usable in unittests
 version (unittest)
 public bool isValid (const ref Block block, Height prev_height,
     Hash prev_hash, scope UTXOFinder findUTXO,
-    size_t active_enrollments, scope PayloadChecker checkPayload) nothrow @safe
+    size_t active_enrollments, scope PayloadChecker checkPayload,
+    ulong enrollment_cycle = 0, string file = __FILE__, size_t line = __LINE__) nothrow @safe
+{
+    string reason = isInvalidReason(block, prev_height, prev_hash, findUTXO,
+        active_enrollments, checkPayload);
+    if (reason !is null)
+        log.trace("[{}:{}] block height {} is invalid: {}",
+            file, line, block.header.height, reason);
+    return reason is null;
+}
+
+/// Ditto but returns `bool`, only usable in unittests
+version (unittest)
+public bool isNotValid (const ref Block block, Height prev_height,
+    Hash prev_hash, scope UTXOFinder findUTXO,
+    size_t active_enrollments, scope PayloadChecker checkPayload,
+    ulong enrollment_cycle = 0) nothrow @safe
 {
-    return isInvalidReason(block, prev_height, prev_hash, findUTXO,
-        active_enrollments, checkPayload) is null;
+    string reason = isInvalidReason(block, prev_height, prev_hash, findUTXO,
+        active_enrollments, checkPayload);
```

Comment:
Did you intend to add a log call here too?

Reply:
Not required in this function as the calling function logs when the test is positive test but fails. I reverted this change so return without temp var.